diff --git a/desmume/src/GPU.cpp b/desmume/src/GPU.cpp
index 84006de..993f452 100644
--- a/desmume/src/GPU.cpp
+++ b/desmume/src/GPU.cpp
@@ -923,10 +923,15 @@ FORCEINLINE void GPU::___setFinalColorBck(u16 color, const size_t srcX, const bo
 	}
 }
 
+#ifdef MSB_FIRST
 //unpacks an _OAM_ structure from the provided oam buffer (should point at OAM 0) and provided OAM index.
 //is endian-safe
 void SlurpOAM(_OAM_ *oam_output, void *oam_buffer, const size_t oam_index)
+#else
+_OAM_ *SlurpOAM(void *oam_buffer, int oam_index)
+#endif
 {
+#ifdef MSB_FIRST
 	const u16 *u16_oam_buffer = (u16 *)oam_buffer;
 	const size_t u16_offset = oam_index << 2;
 	const u16 attr[4]	= {LE_TO_LOCAL_16(u16_oam_buffer[u16_offset + 0]),
@@ -952,6 +957,10 @@ void SlurpOAM(_OAM_ *oam_output, void *oam_buffer, const size_t oam_index)
 	oam_output->PaletteIndex = (attr[2]>>12)&0xF;
 
 	oam_output->attr3 = attr[3];
+#else
+   u16* u16_oam_buffer = (u16*)oam_buffer;
+   return (_OAM_*)&u16_oam_buffer[oam_index << 2];
+#endif
 }
 
 //gets the affine parameter associated with the specified oam index.
@@ -968,11 +977,18 @@ u16 SlurpOAMAffineParam(void *oam_buffer, const size_t oam_index)
 //because we need the speed, and because it is inaccurate anyway
 static void mosaicSpriteLinePixel(GPU *gpu, const size_t x, u16 l, u16 *dst, u8 *dst_alpha, u8 *typeTab, u8 *prioTab)
 {
+#ifdef MSB_FIRST
 	_OAM_ spriteInfo;
 	SlurpOAM(&spriteInfo,gpu->oam,gpu->sprNum[x]);
 	const bool enabled = spriteInfo.Mosaic!=0;
 	if (!enabled)
 		return;
+#else
+   _OAM_* spriteInfo = SlurpOAM(gpu->oam, gpu->sprNum[x]);
+   if (!spriteInfo->Mosaic)
+      return;
+	const bool enabled = true;
+#endif
 
 	const bool opaque = prioTab[x] <= 4;
 
@@ -1573,9 +1589,13 @@ void GPU::_spriteRender(u16 *dst, u8 *dst_alpha, u8 *typeTab, u8 *prioTab)
 
 	for (size_t i = 0; i < 128; i++)
 	{
+#ifdef MSB_FIRST
 		_OAM_ theOAM;
 		_OAM_*spriteInfo = &theOAM;
 		SlurpOAM(spriteInfo, gpu->oam, i);
+#else
+      _OAM_* spriteInfo = SlurpOAM(gpu->oam, i);
+#endif
 
 		//for each sprite:
 		if (cost >= 2130)
diff --git a/desmume/src/GPU.h b/desmume/src/GPU.h
index 1c82f10..c4955cb 100644
--- a/desmume/src/GPU.h
+++ b/desmume/src/GPU.h
@@ -525,6 +525,7 @@ enum GPU_OBJ_MODE
 	GPU_OBJ_MODE_Bitmap = 3
 };
 
+#ifdef MSB_FIRST
 struct _OAM_
 {
 	//attr0
@@ -548,6 +549,39 @@ struct _OAM_
 };
 
 void SlurpOAM(_OAM_ *oam_output, void *oam_buffer, const size_t oam_index);
+#else
+union _OAM_tag
+{
+   u16 attr[4];
+
+   struct
+   {
+      //attr0
+      unsigned Y:8;
+      unsigned RotScale:2;
+      unsigned Mode:2;
+      unsigned Mosaic:1;
+      unsigned Depth:1;
+      unsigned Shape:2;
+      //att1
+      signed X:9;
+      unsigned RotScalIndex:3;
+      unsigned HFlip:1;
+      unsigned VFlip:1;
+      unsigned Size:2;
+      //attr2
+      unsigned TileIndex:10;
+      unsigned Priority:2;
+      unsigned PaletteIndex:4;
+      //attr3
+   	unsigned attr3:16;
+   };
+};
+
+typedef const _OAM_tag _OAM_;
+
+_OAM_* SlurpOAM(void* oam_buffer, int oam_index);
+#endif
 u16 SlurpOAMAffineParam(void *oam_buffer, const size_t oam_index);
 
 typedef struct
